File: code/classes/protocol.m
#import &lt;Foundation/Foundation.h&gt;

// Protocol in Objective-C
@protocol Drawable &lt;NSObject&gt;
- (void)draw;
@end

@interface Circle : NSObject &lt;Drawable&gt;
@end

@implementation Circle
- (void)draw {
    NSLog(@"Drawing a circle");
}
@end

int main() {
    Circle *circle = [[Circle alloc] init];
    [circle draw];
    return 0;
}

File: code/classes/protocol.c
#include &lt;stdio.h&gt;

// Protocol in C
// (simulated with a struct of function pointers)
typedef struct {
    void (*draw)(void);
} Drawable;

// Circle "implements" Drawable
void circle_draw(void) {
    printf("Drawing a circle\n");
}

void draw_shape(Drawable *d) {
    d-&gt;draw();
}

int main() {
    Drawable circle = {circle_draw};
    draw_shape(&circle);
    return 0;
}
